### DISCUSSION TITLE
Modifying collections

### DISCUSSION THREAD

**Discussion by cfjedimaster** [11ty/eleventy#2045] [Uncategorized]
On my site, I create a collection via glob:

```
   let posts = collection.getFilteredByGlob("_posts/**/*.md");
```

I then loop over them all and tweak the permalink and outputPath. This was done to ensure my URLs didn't change when I first migrated to Eleventy.

I'm testing something (not really relevant right now) and I wanted to see what would happen if I *removed* items from a collection. I built my logic, and then used `posts.splice(i,1)` to remove items. I console.log the size and could confirm that the right number of items were removed, but Eleventy still wrote out each and every post.

So my question is - we're allowed to modify collection data, but not remove? 

On the flip side, what about adding? If we can't remove, I could perhaps use two collections. The initial collection based on glob, and a second collection where I add one by one. Is that allowed?

**Comment by nhoizey** [comment]
A collection is "just" a list of things. Such things can be Markdown files, with your usage of `getFilteredByGlob()`.

But even if you don't define this specific collection, any template file will be transformed into another file (often HTML) in the build folder.

So removing items from your collection with `splice()` doesn't actually remove the source Markdown files.

If you don't want a source file to become a built file, its `permalink` should have a `false` value.

Additionally, you can indeed have any number of collections.

For example:

```
let allMdFiles = collection.getFilteredByGlob("**/*.md");
let allPosts = collection.getFilteredByGlob("_posts/**/*.md");
let allPostsFrom2021 = collection.getFilteredByGlob("_posts/2021/**/*.md"); // if you have a folder per year
```

**Reply by cfjedimaster** [reply]
So most of this is duh - by that I mean, I was being dense. ;) Some follow ups though:

If I set permalink to false, are you saying it won't be saved? I'm seeing it be saved still. Here's my code (pardon the mess...)

```
eleventyConfig.addCollection("posts", collection => {
   let posts = collection.getFilteredByGlob("_posts/**/*.md");

   for(let i=posts.length-1;i>=0;i--) {
      // hide old posts
      if(posts[i].outputPath.includes('200')) {
        //posts.splice(i, 1);
        posts[i].data.permalink = false;
      } else {
        posts[i].data.permalink += '.html';
        posts[i].outputPath += '/index.html';
      }
    }

	return posts;
});
```

To this: "Additionally, you can indeed have any number of collections."  I know we can have any number of collections, but can we create a collection over N steps? So can I make an empty collection, loop, and add a document one by one?

**Reply by pdehaan** [reply]
Disclaimer, I have no idea what I'm doing or was trying to accomplish, but this was mildly interesting to me and possibly useful for my own [selfish] future reference.

```js
// ./src/posts/posts.11tydata.js
module.exports = {
  eleventyComputed: {
    permalink(data) {
      if (data.page.date.getFullYear() < 2020) {
        return false;
      }
      return data.permalink;
    }
  }
};
```

```sh
$ tree src

src
└── posts
    ├── 2001/11/23/index.md
    ├── 2009/12/15/index.md
    ├── 2019/10/30/index.md
    ├── 2020/12/31/index.md
    ├── 2021/10/22/index.md
    ├── index.md
    └── posts.11tydata.js

16 directories, 7 files
```

And now when I build my fake blog, it ignores the posts from 2001, 2009, and 2019.

```sh
$ npm run build

> 11ty-collection-filter@1.0.0 build /private/tmp/11ty-collection-filter
> eleventy

Writing www/posts/index.html from ./src/posts/index.md.
Writing www/posts/2020/12/31/index.html from ./src/posts/2020/12/31/index.md.
Writing www/posts/2021/10/22/index.html from ./src/posts/2021/10/22/index.md.
Wrote 3 files in 0.13 seconds (v0.12.1)
```

```sh
$ tree www

www
└── posts
    ├── 2020/12/31/index.html
    ├── 2021/10/22/index.html
    └── index.html

7 directories, 3 files
```

And all my templates set the post date in the front matter since 11ty doesn't convert /2021/10/22/ into a date (but I believe does if I would have named it "2021-10-22-some-blog-title-slug-here.md"):

```md
---
title: "Post from 2020"
date: "2020-12-31T08:08:00+06:00"
categories: [coldfusion]
---

# {{ title }}

{{ page.date }}
```

**Reply by cfjedimaster** [reply]
Oh fascinating - I'll give this a try. It may be a few days. FYI, the idea I'm toying with is trying to get Eleventy to build 2010 and later blog posts, and serve up 2003-2010 via the new serverless feature. My thinking is that it might make my builds *much* quicker in production. (In case I don't get a chance to return to this soon, thank you.)

**Reply by pdehaan** [reply]
Free business idea: Blog only displays the last 100 posts. The other 5900+ are available as an in app purchase in PDF format.

But yes. Keep us updated. I haven't played w/ the new serverless stuff yet so am quite interested how it all works out for you (especially since you have a rather large site with oodles of content from the past 18 years, which is a pretty cool data set).
